web-services/query/src/main/java/datawave/webservice/query/logic/composite/CompositeQueryLogic.java

```diff
@@ -145,52 +164,80 @@ public void run() {
     private volatile boolean interrupted = false;
     private volatile CountDownLatch startLatch = null;
     private volatile CountDownLatch completionLatch = null;
-    private List<Entry<QueryLogic<?>,QueryLogicHolder>> logicState = new ArrayList<>();
+    private Map<String,QueryLogicHolder> logicState = new HashMap<>();
     private volatile CompositeQueryLogicResults results = null;
     
     public CompositeQueryLogic() {}
     
     public CompositeQueryLogic(CompositeQueryLogic other) {
         super(other);
         if (other.queryLogics != null) {
-            this.queryLogics = new ArrayList<>();
-            for (QueryLogic<?> logic : other.queryLogics) {
+            this.queryLogics = new HashMap<>();
+            for (Map.Entry<String,QueryLogic<?>> entry : other.queryLogics.entrySet()) {
                 try {
-                    logic = (QueryLogic) logic.clone();
+                    this.queryLogics.put(entry.getKey(), (QueryLogic) entry.getValue().clone());
                 } catch (CloneNotSupportedException e) {
                     throw new RuntimeException(e);
                 }
-                this.queryLogics.add(logic);
             }
             setPrincipal(other.getPrincipal());
         }
         this.allMustInitialize = other.allMustInitialize;
     }
     
+    public static Set<Authorizations> updateRuntimeAuthorizationsAndQueryAuths(QueryLogic<?> logic, Query settings) throws AuthorizationException {
+        Set<String> requestedAuths = new HashSet<>(AuthorizationsUtil.splitAuths(settings.getQueryAuthorizations()));
+        
+        // determine the valid authorizations for this call to be the user's auths for this logic
+        DatawavePrincipal principal = (DatawavePrincipal) logic.getPrincipal();
+        if (logic.getUserOperations() != null) {
+            principal = logic.getUserOperations().getRemoteUser(principal);
```

Comment:
See question elsewhere regarding the implications of replacing the local principal with the remote principal.